upstairs/src/block_req.rs

```diff
@@ -11,41 +11,34 @@ use tokio::sync::oneshot;
 #[derive(Debug)]
 pub(crate) struct BlockReq {
     pub op: BlockOp,
-    sender: oneshot::Sender<Result<(), CrucibleError>>,
+    pub res: BlockRes,
 }
 
-impl BlockReq {
-    pub fn new(
-        op: BlockOp,
-        sender: oneshot::Sender<Result<(), CrucibleError>>,
-    ) -> BlockReq {
-        Self { op, sender }
-    }
-
-    /// Return a copy of the block op
-    pub fn op(&self) -> BlockOp {
-        self.op.clone()
-    }
-
-    /// Consume this BlockReq and send Ok to the receiver
+#[must_use]
+#[derive(Debug)]
+pub(crate) struct BlockRes(Option<oneshot::Sender<Result<(), CrucibleError>>>);
+impl BlockRes {
+    /// Consume this BlockRes and send Ok to the receiver
     pub fn send_ok(self) {
         self.send_result(Ok(()))
     }
 
-    /// Consume this BlockReq and send an Err to the receiver
+    /// Consume this BlockRes and send an Err to the receiver
     pub fn send_err(self, e: CrucibleError) {
         self.send_result(Err(e))
     }
 
-    /// Consume this BlockReq and send a Result to the receiver
-    pub fn send_result(self, r: Result<(), CrucibleError>) {
+    /// Consume this BlockRes and send a Result to the receiver
+    fn send_result(mut self, r: Result<(), CrucibleError>) {
         // XXX this eats the result!
-        let _ = self.sender.send(r);
+        let _ = self.0.take().expect("sender was populated").send(r);
     }
-
-    /// Consume this BlockReq and return the inner oneshot sender
-    pub fn take_sender(self) -> oneshot::Sender<Result<(), CrucibleError>> {
-        self.sender
+}
+impl Drop for BlockRes {
+    fn drop(&mut self) {
+        // Dropping a BlockRes without issuing a completion would mean the
+        // associated waiter would be stuck waiting forever for a result.
```

Comment:
I like this check! But: instead of asserting, would a `send_err` be appropriate?

Reply:
It could be.  I'll leave that decision up to y'all.  Not dealing with a `BlockRes` before it's dropped is definitely a programmer error (vs operational error), but since it's not being enforced by the compiler, I can appreciate not wanting an assertion.  That said, having an assert helped me track down where such drops were occurring.

Reply:
Some additional context: propolis has been using a similar pattern for its [block Request](https://github.com/oxidecomputer/propolis/blob/85dc6db36709981c1d35b0651a18e20caee0b73f/lib/propolis/src/block/mod.rs#L176-L182) abstraction.

Reply:
Sorry for my ignorance here,
Who is dropping the `BlockRes` in this situation?  Is it the crucible upstairs side that would
have done the drop?

Reply:
Yeah, it would be upstairs which could drop a `BlockRes` rather than properly "disposing" of it by calling `send_ok` or `send_err`.

Reply:
I like panic for programmer errors (and, yeah, I benefit so often by catching myself).